### PR TITLE
Bug 1926522: oc adm catalog mirror tmp directory cleanup

### PR DESCRIPTION
**Pull request by tylerslaton**
WIP: This PR supersedes [PR 741](https://github.com/openshift/oc/pull/741).

### CODE REVIEW

File: pkg/cli/admin/catalog/mirror.go [hunk 1]
```diff
@@ -585,6 +590,20 @@ func aggregateICSPs(icsps [][]byte) []byte {
...
+	if o.TempDir {
+		dir := strings.Split(o.IndexPath, ":")
+		if err := os.RemoveAll(dir[1]); err != nil {
+			return fmt.Errorf("unable to delete dir %s", dir[1])
+		}
+		fmt.Fprintf(o.IOStreams.Out, "deleted dir %s\n", dir[1])
+		return nil
+	}
+	return nil
+}
```
Comment (njhale) [inline pkg/cli/admin/catalog/mirror.go:605]: nit: return early and save on indentation

```go
if !o.TempDir {
    return nil
}
dir := strings.Split(o.IndexPath, ":")
//...
```

File: pkg/cli/admin/catalog/mirror.go [hunk 2]
```diff
@@ -585,6 +590,20 @@ func aggregateICSPs(icsps [][]byte) []byte {
 	return aggregation
 }
 
+func (o *MirrorCatalogOptions) PostRun() error {
+	// If we have NOT set --path, the TempDir is set to true and we will delete
+	// Temporary folder.
+	if o.TempDir {
+		dir := strings.Split(o.IndexPath, ":")
+		if err := os.RemoveAll(dir[1]); err != nil {
+			return fmt.Errorf("unable to delete dir %s", dir[1])
```
Comment (njhale) [inline pkg/cli/admin/catalog/mirror.go:599]: Wrap errors with `%w` to bubble up:

```go
return fmt.Errorf("unable to delete dir %s: %w", dir[1], err)
```

File: pkg/cli/admin/catalog/mirror.go [hunk 3]
```diff
@@ -585,6 +590,20 @@ func aggregateICSPs(icsps [][]byte) []byte {
 	return aggregation
 }
 
+func (o *MirrorCatalogOptions) PostRun() error {
+	// If we have NOT set --path, the TempDir is set to true and we will delete
+	// Temporary folder.
+	if o.TempDir {
+		dir := strings.Split(o.IndexPath, ":")
```
Comment (njhale) [inline pkg/cli/admin/catalog/mirror.go:597]: I find it always helps to be defensive and check the length of a slice before directly accessing an index. The assumption that `TempDir` equates to `len(dir) > 1` may not always hold.

Alternatively, you could encapsulate the assumption in a named type; e.g. `type DirPath string` that has a `TempDir() string` method, make `MirrorCatalogOptions.IndexPath` of that type, and call `tmp := o.IndexPath.TempDir()` here.

File: pkg/cli/admin/catalog/mirror.go [hunk 4]
```diff
@@ -585,6 +590,20 @@ func aggregateICSPs(icsps [][]byte) []byte {
 	return aggregation
 }
 
+func (o *MirrorCatalogOptions) PostRun() error {
```
Comment (njhale) [inline pkg/cli/admin/catalog/mirror.go:593]: Does this method need to be exported anymore?

**Review by njhale — COMMENTED** [review]
Looking good so far!

Added a few comments/questions.

File: pkg/cli/admin/catalog/mirror.go (hunk 4 above)
Comment (tylerslaton) [reply pkg/cli/admin/catalog/mirror.go:593]: Nope, good point.

File: pkg/cli/admin/catalog/mirror.go (hunk 3 above)
Comment (tylerslaton) [reply pkg/cli/admin/catalog/mirror.go]: Would the:
```go
if err := os.RemoveAll(dir[1]); err != nil {
```
Not already be catching this? Maybe I'm misunderstanding.

Comment (tylerslaton) [reply pkg/cli/admin/catalog/mirror.go]: How does something like this look to you?

```go
dir := strings.Split(o.IndexPath, ":")
if len(dir) != 1 {
    return fmt.Errorf("could not find dir to cleanup")
}
```

**Comment by njhale** [comment]
Changes addressed the holdable issue.

/hold cancel
/lgtm

**Review by tbuskey — COMMENTED** [review]
/lgtm

File: pkg/cli/admin/catalog/mirror.go [hunk 5]
```diff
@@ -585,6 +594,27 @@ func aggregateICSPs(icsps [][]byte) []byte {
...
+func (o *MirrorCatalogOptions) postRun() error {
+	// If we have NOT set --path, the TempDir is set to true and we will delete
+	// Temporary folder.
+	if !o.TempDir {
+		return nil
+	}
+
+	dir := strings.Split(o.IndexPath, ":")
+	if len(dir) < 2 {
+		return fmt.Errorf("error cleaning up temp dir: could not find dir")
```
Comment (ankitathomas) [inline pkg/cli/admin/catalog/mirror.go:606]: nit: possibly useful to print the invalid o.IndexPath

**Review by ankitathomas — APPROVED** [review]
/lgtm

**Comment by soltysh** [comment]
/override ci/prow/e2e-agnostic-cmd
this is currently being worked on elsewhere

**Pull request merged**